File: web/src/weatherMapper.js
const BAD_CONDITION_KEYWORDS = [
  'RAIN',
  'SNOW',
  'STORM',
  'SLEET',
  'HAIL',
  'DRIZZLE',
  'THUNDERSTORM',
  'WINDY',
  'FOG',
];

function normalizeString(value) {
  if (typeof value !== 'string') {
    return '';
  }
  return value.trim().toUpperCase();
}

function conditionContainsBadWeather(condition) {
  return BAD_CONDITION_KEYWORDS.some((keyword) => condition.includes(keyword));
}

function conditionContainsRainOrSnow(condition) {
  return condition.includes('RAIN') || condition.includes('SNOW');
}

export function mapGoogleWeatherToIsBad(condition, precipType, precipProbability) {
  const normalizedCondition = normalizeString(condition);
  const normalizedPrecipType = normalizeString(precipType);

  if (normalizedPrecipType === 'RAIN' || normalizedPrecipType === 'SNOW' || normalizedPrecipType === 'MIX' || normalizedPrecipType === 'SLEET') {
    return true;
  }

  if (normalizedPrecipType === 'NONE') {
    return conditionContainsRainOrSnow(normalizedCondition);
  }

  if (conditionContainsBadWeather(normalizedCondition)) {
    return true;
  }

  if (typeof precipProbability === 'number' && precipProbability >= 50) {
    return true;
  }

  return false;
}

File: android/app/src/main/java/com/commuteoptimizer/widget/service/WeatherMapper.kt
package com.commuteoptimizer.widget

private val BAD_CONDITION_KEYWORDS = listOf(
    "RAIN",
    "SNOW",
    "STORM",
    "SLEET",
    "HAIL",
    "DRIZZLE",
    "THUNDERSTORM",
)

private fun normalizeWeatherValue(value: String?): String {
    return value?.trim()?.uppercase() ?: ""
}

private fun containsBadConditionKeyword(condition: String): Boolean {
    return BAD_CONDITION_KEYWORDS.any { keyword -> condition.contains(keyword) }
}

fun mapGoogleWeatherToIsBad(
    condition: String?,
    precipitationProbability: Int? = null,
    precipitationType: String? = null
): Boolean {
    val normalizedCondition = normalizeWeatherValue(condition)
    val normalizedPrecipitationType = normalizeWeatherValue(precipitationType)

    if (normalizedPrecipitationType == "RAIN" || normalizedPrecipitationType == "SNOW" || normalizedPrecipitationType == "MIX" || normalizedPrecipitationType == "SLEET") {
        return true
    }

    if (normalizedPrecipitationType == "NONE") {
        return false
    }

    if (containsBadConditionKeyword(normalizedCondition)) {
        return true
    }

    if ((precipitationProbability ?: -1) >= 50) {
        return true
    }

    return false
}
